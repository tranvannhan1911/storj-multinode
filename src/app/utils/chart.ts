// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BandwidthRollup } from '@/bandwidth';
import { SizeBreakpoints } from '@/private/memory/size';
import { Stamp } from '@/storage';

const shortMonthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec'];

/**
 * Used to display correct and convenient data on chart.
 */
export class Chart {
    /**
     * Brings chart data to a more compact form.
     * @param data - holds array of chart data in numeric form
     * @returns data - numeric array of normalized data
     */
    public static normalizeChartData(data: number[]): number[] {
        const maxBytes = Math.ceil(Math.max(...data));

        let divider: number = SizeBreakpoints.PB;

        switch (true) {
        case maxBytes < SizeBreakpoints.MB:
            divider = SizeBreakpoints.KB;
            break;
        case maxBytes < SizeBreakpoints.GB:
            divider = SizeBreakpoints.MB;
            break;
        case maxBytes < SizeBreakpoints.TB:
            divider = SizeBreakpoints.GB;
            break;
        case maxBytes < SizeBreakpoints.PB:
            divider = SizeBreakpoints.TB;
            break;
        }

        return data.map(elem => elem / divider);
    }

    /**
     * gets chart data dimension depending on data size.
     * @param data - holds array of chart data in numeric form
     * @returns dataDimension - string of data dimension
     */
    public static getChartDataDimension(data: number[]): string {
        const maxBytes = Math.ceil(Math.max(...data));

        let dataDimension: string;

        switch (true) {
        case maxBytes < SizeBreakpoints.MB:
            dataDimension = 'KB';
            break;
        case maxBytes < SizeBreakpoints.GB:
            dataDimension = 'MB';
            break;
        case maxBytes < SizeBreakpoints.TB:
            dataDimension = 'GB';
            break;
        case maxBytes < SizeBreakpoints.PB:
            dataDimension = 'TB';
            break;
        default:
            dataDimension = 'PB';
        }

        return dataDimension;
    }

    /**
     * Used to display correct number of days on chart's labels.
     *
     * @returns daysDisplayed - array of days converted to a string by using the current or specified locale
     */
    public static daysDisplayedOnChart(): string[] {
        const daysDisplayed = Array<string>(new Date().getUTCDate());
        const currentMonth = shortMonthNames[new Date().getUTCMonth()].toUpperCase();

        for (let i = 0; i < daysDisplayed.length; i++) {
            daysDisplayed[i] = `${currentMonth} ${i + 1}`;
        }

        if (daysDisplayed.length === 1) {
            daysDisplayed.unshift('0');
        }

        return daysDisplayed;
    }

    /**
     * Adds missing bandwidth usage for bandwidth chart data for each day of month.
     * @param fetchedData - array of data that is spread over missing bandwidth usage for each day of the month
     * @returns bandwidthChartData - array of filled data
     */
    public static populateEmptyBandwidth(fetchedData: BandwidthRollup[]): BandwidthRollup[] {
        const bandwidthChartData: BandwidthRollup[] = new Array(new Date().getUTCDate());
        const data: BandwidthRollup[] = fetchedData || [];

        if (data.length === 0) {
            return bandwidthChartData;
        }

        outer:
        for (let i = 0; i < bandwidthChartData.length; i++) {
            const date = i + 1;

            for (let j = 0; j < data.length; j++) {
                if (data[j].intervalStart.getUTCDate() === date) {
                    bandwidthChartData[i] = data[j];
                    continue outer;
                }
            }

            bandwidthChartData[i] = BandwidthRollup.emptyWithDate(date);
        }

        if (bandwidthChartData.length === 1) {
            bandwidthChartData.unshift(BandwidthRollup.emptyWithDate(1));
            bandwidthChartData[0].intervalStart.setUTCHours(0, 0, 0, 0);
        }

        return bandwidthChartData;
    }

    /**
     * Adds missing stamps for storage chart data for each day of month.
     * @param fetchedData - array of data that is spread over missing stamps for each day of the month
     * @returns storageChartData - array of filled data
     */
    public static populateEmptyStamps(fetchedData: Stamp[]): Stamp[] {
        const storageChartData: Stamp[] = new Array(new Date().getUTCDate());
        const data: Stamp[] = fetchedData || [];

        if (data.length === 0) {
            return storageChartData;
        }

        outer:
        for (let i = 0; i < storageChartData.length; i++) {
            const date = i + 1;

            for (let j = 0; j < data.length; j++) {
                if (data[j].intervalStart.getUTCDate() === date) {
                    storageChartData[i] = data[j];
                    continue outer;
                }
            }

            storageChartData[i] = Stamp.emptyWithDate(date);
        }

        if (storageChartData.length === 1) {
            storageChartData.unshift(Stamp.emptyWithDate(1));
            storageChartData[0].intervalStart.setUTCHours(0, 0, 0, 0);
        }

        return storageChartData;
    }
}
