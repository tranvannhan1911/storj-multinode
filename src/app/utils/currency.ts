// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Size class contains currency related functionality such as convertation.
 */
export class Currency {
    /**
     * dollarsFromCents converts cents to dollars with prefix.
     * @param cents count
     */
    public static dollarsFromCents(cents: number): string {
        return `$${(cents / 100).toFixed(2)}`;
    }
}
