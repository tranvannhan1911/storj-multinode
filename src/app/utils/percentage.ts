// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Percentage class contains percentage related functionality.
 */
export class Percentage {
    /**
     * dollarsFromCents converts cents to dollars with prefix.
     * @param number float
     */
    public static fromFloat(number: number): string {
        return `${parseFloat(`${(number * 100).toFixed(1)}`)}%`;
    }
}
