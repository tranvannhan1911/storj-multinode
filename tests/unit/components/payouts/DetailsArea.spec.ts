// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import { Currency } from '@/app/utils/currency';

import DetailsArea from '@/app/components/payouts/DetailsArea.vue';

const localVue = createLocalVue();

localVue.use(Vuex);

localVue.filter('centsToDollars', (cents: number): string => Currency.dollarsFromCents(cents));

describe('DetailsArea', (): void => {
    it('renders correctly', (): void => {
        const wrapper = shallowMount(DetailsArea, {
            localVue,
            propsData: {
                totalEarned: 5000,
                totalPaid: 60000,
                totalHeld: 700,
                period: 'April, 2021',
            },
        });

        expect(wrapper).toMatchSnapshot();
    });
});
