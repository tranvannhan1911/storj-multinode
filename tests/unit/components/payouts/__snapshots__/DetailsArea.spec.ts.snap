// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsArea renders correctly 1`] = `
<info-block-stub>
  <div class="details-area">
    <header class="details-area__header">
      <h3 class="details-area__header__title">Details</h3>
      <p class="details-area__header__period">April, 2021</p>
    </header>
    <div class="details-area__content">
      <div class="details-area__content__item">
        <p class="details-area__content__item__label">EARNED</p>
        <p class="details-area__content__item__value">$50.00</p>
      </div>
      <div class="details-area__content__item">
        <p class="details-area__content__item__label">HELD</p>
        <p class="details-area__content__item__value">$7.00</p>
      </div>
      <div class="details-area__content__item">
        <p class="details-area__content__item__label">PAID</p>
        <p class="details-area__content__item__value">$600.00</p>
      </div>
    </div>
  </div>
</info-block-stub>
`;
